tests/bsim/bluetooth/host/att/long_read/testlib/att_read.c

```diff
@@ -0,0 +1,390 @@
+#include <zephyr/kernel.h>
+#include <zephyr/sys/__assert.h>
+
+#include "att_read.h"
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(att_read, LOG_LEVEL_DBG);
+
+struct bt_testlib_att_read_closure {
+	uint8_t att_err;
+	struct bt_conn *conn;
+	struct bt_gatt_read_params params;
+	uint16_t *result_size;
+	uint16_t *result_handle;
+	struct net_buf_simple *result_data;
+	struct k_mutex lock;
+	struct k_condvar done;
+	uint16_t *att_mtu;
+	bool long_read;
+};
+
+static inline uint8_t att_read_cb(struct bt_conn *conn, uint8_t att_err,
+				  struct bt_gatt_read_params *params, const void *read_data,
+				  uint16_t read_len)
+{
+	struct bt_testlib_att_read_closure *ctx =
+		CONTAINER_OF(params, struct bt_testlib_att_read_closure, params);
+
+	k_mutex_lock(&ctx->lock, K_FOREVER);
+
+	if (read_data == NULL) {
+		__ASSERT_NO_MSG(ctx->long_read);
+		k_condvar_signal(&ctx->done);
+		k_mutex_unlock(&ctx->lock);
+		return BT_GATT_ITER_STOP;
+	}
+
+	ctx->att_err = att_err;
+
+	if (!att_err && ctx->result_handle) {
+		*ctx->result_handle = params->by_uuid.start_handle;
+	}
+
+	if (!att_err && ctx->result_size) {
+		LOG_DBG("Adding %u bytes to result", read_len);
+		*ctx->result_size += read_len;
+		if (*ctx->result_size > 512) {
+			LOG_ERR("result_size > 512");
+		}
+	}
+
+	if (!att_err && ctx->result_data) {
+		uint16_t result_data_size =
+			MIN(read_len, net_buf_simple_tailroom(ctx->result_data));
+		net_buf_simple_add_mem(ctx->result_data, read_data, result_data_size);
+	}
+
+	if (!att_err && ctx->att_mtu) {
+		*ctx->att_mtu = params->_att_mtu;
+	}
+
+	if (ctx->long_read) {
+		/* Don't signal `&ctx->done` */
+		k_mutex_unlock(&ctx->lock);
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	k_condvar_signal(&ctx->done);
+	k_mutex_unlock(&ctx->lock);
+	return BT_GATT_ITER_STOP;
+}
+
+static inline int bt_testlib_sync_bt_gatt_read(struct bt_testlib_att_read_closure *ctx)
+{
+	int api_err;
+
+	ctx->params.func = att_read_cb;
+
+	k_mutex_init(&ctx->lock);
+	k_condvar_init(&ctx->done);
+
+	k_mutex_lock(&ctx->lock, K_FOREVER);
+
+	api_err = bt_gatt_read(ctx->conn, &ctx->params);
+
+	if (!api_err) {
+		k_condvar_wait(&ctx->done, &ctx->lock, K_FOREVER);
+	}
+
+	k_mutex_unlock(&ctx->lock);
+
+	if (api_err) {
+		__ASSERT_NO_MSG(api_err < 0);
+		return api_err;
+	}
+
+	__ASSERT_NO_MSG(ctx->att_err >= 0);
+	return ctx->att_err;
+}
+
+int bt_testlib_att_read_by_type_sync(struct net_buf_simple *result_data, uint16_t *result_size,
+				     uint16_t *result_handle, uint16_t *result_att_mtu,
+				     struct bt_conn *conn, enum bt_att_chan_opt bearer,
+				     const struct bt_uuid *type, uint16_t start_handle,
+				     uint16_t end_handle)
+{
+	struct bt_testlib_att_read_closure ctx = {
+		.result_handle = result_handle,
+		.result_size = result_size,
+		.conn = conn,
+		.result_data = result_data,
+		.att_mtu = result_att_mtu,
+		.params = {
+			.by_uuid = {.uuid = type,
+				    .start_handle = start_handle,
+				    .end_handle = end_handle},
```

Comment:
It's you against clang-format + checkpatch.

Reply:
> It's you against clang-format + checkpatch. With the comma there, clang-format makes checkpatch whine.

Why is Checkpath whining? We do similar formatting several other places (e.g. 
```c
	static struct bt_vcp_vol_ctlr_cb vcp_cbs = {
		.discover = vcs_discover_cb,
		.vol_down = vcs_write_cb,
		.vol_up = vcs_write_cb,
		.mute = vcs_write_cb,
		.unmute = vcs_write_cb,
		.vol_down_unmute = vcs_write_cb,
		.vol_up_unmute = vcs_write_cb,
		.vol_set = vcs_write_cb,
		.state = vcs_state_cb,
		.flags = vcs_flags_cb,
		.vocs_cb = {
			.state = vocs_state_cb,
			.location = vocs_location_cb,
			.description = vocs_description_cb,
			.set_offset = vocs_write_cb,
		},
```

But if it can't pass CI, then it is OK to leave as is.

Reply:
My clang-format does this to your code:
```c
	static struct bt_vcp_vol_ctlr_cb vcp_cbs = {
		.discover = vcs_discover_cb,
		.vol_down = vcs_write_cb,
		.vol_up = vcs_write_cb,
		.mute = vcs_write_cb,
		.unmute = vcs_write_cb,
		.vol_down_unmute = vcs_write_cb,
		.vol_up_unmute = vcs_write_cb,
		.vol_set = vcs_write_cb,
		.state = vcs_state_cb,
		.flags = vcs_flags_cb,
		.vocs_cb =
			{
				.state = vocs_state_cb,
				.location = vocs_location_cb,
				.description = vocs_description_cb,
				.set_offset = vocs_write_cb,
			},
	};
```
It's different. Do I gave a misconfiguration in my editor?

Reply:
By properly inserting commas after the last field, I get
```c
static void foo(void)
{
	struct bt_testlib_att_read_closure ctx = {
		.result_handle = result_handle,
		.result_size = result_size,
		.conn = conn,
		.result_data = result_data,
		.att_mtu = result_att_mtu,
		.params =
			{
				.by_uuid =
					{
						.uuid = type,
						.start_handle = start_handle,
						.end_handle = end_handle,
					},
				IF_ENABLED(CONFIG_BT_EATT, (.chan_opt = bearer)),
			},
	};
}
```

Reply:
> It's different. What does your clang-format produce?

Yes, I see that clang-format will put the 2nd curly bracket on a new line, which is different. However, the point of that example was that checkpath shouldn't complain about it.

Reply:
What is checkpath complaining about? Is it related to the `IF_ENABLED(CONFIG_BT_EATT, (.chan_opt = bearer)),` perhaps?

Reply:
It will complain that the curly brace is on a new line.

Reply:
I see. 

Then leave it as is, and then we can look into modifying the clang-format config so that it doesn't do that (that always annoyed me anyways).

Reply:
Will you create a ticket? I think we need autoformatting for the issues you point out other places in this PR. (Like requiring empty lines after code blocks.) Manually commenting and then the author manually fixing is really inefficient use of time.

Reply:
> Manually commenting and then the author manually fixing is really inefficient use of time.

Agreed.

> Will you create a ticket?

Let me check the clang-format docs to see if it has an option for it, and if so, then I'll probably just create a PR to add it.